### PR TITLE
Rename rack to isolation group

### PR DESCRIPTION
**Pull request by cw9**
@xichen2020 @robskillington

### CODE REVIEW

File: placement/algo/sharded_helper.go [hunk 1]
```diff
@@ -88,15 +88,15 @@ type PlacementHelper interface {
 	// Instances returns the list of instances managed by the PlacementHelper.
 	Instances() []placement.Instance
 
-	// HasRackConflict checks if the rack constraint is violated when moving the shard to the target rack.
-	HasRackConflict(shard uint32, from placement.Instance, toRack string) bool
+	// CanMove checks if the shard can be moved from the instance to the target isolation group.
+	CanMove(shard uint32, from placement.Instance, isolationGroup string) bool
```
Comment (robskillington) [inline placement/algo/sharded_helper.go]: Isn't this method the inverse of `CanMove`?  It seems to only return true if you cannot move it (i.e. there is a conflict).

Would a better name be `CannotPlaceShardInIsolationGroupFromInstance(...)`?  Also using parameters `...(shard, isolationGroup, placement.Instance)` might make more sense to match the order of things described by the name.

File: placement/selector/mirrored.go [hunk 2]
```diff
@@ -419,42 +419,42 @@ func (h host) addInstance(port uint32, instance placement.Instance) error {
...
+		return fmt.Errorf("could not add instance %s to host %s, isolation group mismatch: %s and %s",
+			instance.ID(), h.name, instance.IsolationGroup(), h.ig)
 	}
 	h.portToInstance[port] = instance
 	return nil
 }
 
-type rack struct {
-	rack  string
+type ig struct {
```
Comment (robskillington) [inline placement/selector/mirrored.go]: Yeah we should probably call these `type isoGroup struct` or something and put a comment above it.

File: placement/selector/mirrored.go [hunk 3]
```diff
@@ -419,42 +419,42 @@ func (h host) addInstance(port uint32, instance placement.Instance) error {
...
+	ig    string
 	hosts []host
 }
 
-type racksByNumHost []*rack
+type igsByNumHost []*ig
```
Comment (robskillington) [inline placement/selector/mirrored.go]: Similar here `type isoGroupsByNumHost []*isoGroup`?

File: placement/algo/sharded_helper.go (hunk 1 above)
Comment (cw9) [reply placement/algo/sharded_helper.go]: Yeah I think making it `CanMoveShard` is probably more clear, I'm going to reverse the implementation and usage instead

File: generated/proto/placementpb/placement.proto [hunk 4]
```diff
@@ -41,7 +41,7 @@ message Placement {
 
 message Instance {
   string id = 1;
-  string rack = 2;
+  string isolation_group = 2;
```
Comment (xichen2020) [inline generated/proto/placementpb/placement.proto]: I think this is backward compatible, but might be worth double checking.

File: placement/algo/sharded_helper.go [hunk 5]
```diff
@@ -88,15 +88,15 @@ type PlacementHelper interface {
...
+	// CanMoveShard checks if the shard can be moved from the instance to the target isolation group.
+	CanMoveShard(shard uint32, fromInstance placement.Instance, toIsolationGroup string) bool
 }
 
 type helper struct {
 	targetLoad         map[string]int
 	shardToInstanceMap map[uint32]map[placement.Instance]struct{}
-	rackToInstancesMap map[string]map[placement.Instance]struct{}
-	rackToWeightMap    map[string]uint32
+	igToInstancesMap   map[string]map[placement.Instance]struct{}
```
Comment (xichen2020) [inline placement/algo/sharded_helper.go]: nit: would it be more clear to call it `groupToInstancesMap`? `ig` is a bit vague IMO.

File: placement/algo/sharded_helper.go [hunk 6]
```diff
@@ -241,11 +241,11 @@ func (ph *helper) scanCurrentLoad() {
 }
 
 func (ph *helper) buildTargetLoad() {
-	overWeightedRack := 0
+	overWeightedIG := 0
```
Comment (xichen2020) [inline placement/algo/sharded_helper.go]: nit: I'd spell out IG here.

File: placement/service/service_test.go [hunk 7]
```diff
@@ -271,7 +272,7 @@ func TestBadAddReplica(t *testing.T) {
 		10, 1)
 	assert.NoError(t, err)
 
-	// not enough racks/instances
+	// not enough isolation groups/instances
```
Comment (xichen2020) [inline placement/service/service_test.go]: nit: comment should start w/ Upper case letter.

File: placement/service/service_test.go [hunk 8]
```diff
@@ -320,7 +321,7 @@ func TestBadRemoveInstance(t *testing.T) {
 	_, err = p.RemoveInstances([]string{"not_exist"})
 	assert.Error(t, err)
 
-	// not enough racks/instances after removal
+	// not enough isolation groups/instances after removal
```
Comment (xichen2020) [inline placement/service/service_test.go]: nit: comment should start w/ Upper case letter.

File: placement/service/service_test.go [hunk 9]
```diff
@@ -353,7 +354,7 @@ func TestBadReplaceInstance(t *testing.T) {
 	)
 	assert.Error(t, err)
 
-	// not enough rack after replace
+	// not enough isolation groups after replace
```
Comment (xichen2020) [inline placement/service/service_test.go]: nit: comment should start w/ Upper case letter.

File: generated/proto/placementpb/placement.proto (hunk 4 above)
Comment (cw9) [reply generated/proto/placementpb/placement.proto]: Yeah verified this is backward compatible

**Pull request merged**